pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -546,22 +554,61 @@ static boolean isExternalViewConverged(String tableNameWithType,
-      Map<String, Map<String, String>> targetAssignment, int minAvailableReplicas) {
+  /**
+   * Returns the next assignment for the table based on the current assignment and the target assignment with regards to
+   * the minimum available replicas requirement. For strict replica-group mode, track the available instances for all
+   * the segments with the same instances in the next assignment, and ensure the minimum available replicas requirement
+   * is meet. If adding the assignment for a segment breaks the requirement, use the current assignment for the segment.
+   */
+  @VisibleForTesting
+  static Map<String, Map<String, String>> getNextAssignment(Map<String, Map<String, String>> currentAssignment,
+      Map<String, Map<String, String>> targetAssignment, int minAvailableReplicas, boolean enableStrictReplicaGroup) {
     Map<String, Map<String, String>> nextAssignment = new TreeMap<>();
-
-    for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
-      String segmentName = entry.getKey();
-      nextAssignment.put(segmentName,
-          getNextInstanceStateMap(entry.getValue(), targetAssignment.get(segmentName), minAvailableReplicas));
+    if (enableStrictReplicaGroup) {
+      Map<Set<String>, Set<String>> availableInstancesMap = new HashMap<>();
+      for (Map.Entry<String, Map<String, String>> entry : currentAssignment.entrySet()) {
+        String segmentName = entry.getKey();
+        Map<String, String> currentInstanceStateMap = entry.getValue();
+        Map<String, String> targetInstanceStateMap = targetAssignment.get(segmentName);
+        SingleSegmentAssignment assignment =
+            getNextAssignment(currentInstanceStateMap, targetInstanceStateMap, minAvailableReplicas);
+        Set<String> assignedInstances = assignment._instanceStateMap.keySet();
+        Set<String> availableInstances = assignment._availableInstances;
+        Set<String> currentAvailableInstances = availableInstancesMap.get(assignedInstances);
+        if (currentAvailableInstances == null) {
+          // First segment assigned to these instances, use the new assignment and update the available instances
+          nextAssignment.put(segmentName, assignment._instanceStateMap);
+          availableInstancesMap.put(assignedInstances, availableInstances);
+        } else {
+          // There are other segments assigned to the same instances, check the available instances to see if adding the
+          // new assignment can still hold the minimum available replicas requirement
+          availableInstances.retainAll(currentAvailableInstances);
+          if (availableInstances.size() >= minAvailableReplicas) {
+            // New assignment can be added
+            nextAssignment.put(segmentName, assignment._instanceStateMap);
+            availableInstancesMap.put(assignedInstances, availableInstances);
+          } else {
+            // New assignment cannot be added, use the current instance state map
```

Comment:
hmm, does the greedy algorithm always work? Is it possible the iteration order of `currentAssignment.entrySet` matters?

Reply:
It always work. The first segment and all the segments that have the same assignment with the first segment will always be added. We are getting closer to the target assignment in each iteration. In most cases, it is the same number of iterations as the non-strict replica-group one because the collision is rear.

Reply:
Got it.